sbncode/LArRecoProducer/CRUMBS_module.cc

```diff
@@ -0,0 +1,681 @@
+
+    slicePDG = 999999; matchedIndex = 999999;
+    matchedType = "";
+    matchedPurity = -999999.; matchedCompleteness = -999999.;
+  }
+
+  void CRUMBS::SetupMaps(art::Event const& e)
+  {
+
+    unsigned nNu(0), nCos(0);
+
+    if(fProcessNeutrinos)
+      {
+	art::Handle<std::vector<simb::MCTruth> > handleMCTruthNu;
+	e.getByLabel(fGeneratorModuleLabel, handleMCTruthNu);
+	art::FindManyP<simb::MCParticle> truthNuMCPAssn(handleMCTruthNu,e,fMCParticleModuleLabel);
+
+	for (unsigned int i = 0; i < handleMCTruthNu->size(); ++i){
+	  const art::Ptr<simb::MCTruth> mcTruth(handleMCTruthNu, i);
+	  const simb::MCParticle nu = mcTruth->GetNeutrino().Nu();
+
+	  if(!fTpcGeo.InVolume(nu))
+	    fGenTypeMap[i] = "DirtNu";
+	  else
+	    fGenTypeMap[i] = "Nu";
+    
+	  const std::vector<art::Ptr<simb::MCParticle> > particles = truthNuMCPAssn.at(mcTruth.key());
+    
+	  for (auto const& particle : particles)
+	    {
+	      fTrackIDToGenMap[particle->TrackId()] = i;
+	    }
+	  ++nNu;
+	}
+      }
+
+    if(fProcessCosmics)
+      {
+	art::Handle<std::vector<simb::MCTruth> > handleMCTruthCosmic;
+	e.getByLabel(fCosmicModuleLabel, handleMCTruthCosmic);
+
+	art::FindManyP<simb::MCParticle> truthCosmicMCPAssn(handleMCTruthCosmic,e,fMCParticleModuleLabel);
+
+	for (unsigned int i = 0; i < handleMCTruthCosmic->size(); ++i){
+	  const art::Ptr<simb::MCTruth> mcTruth(handleMCTruthCosmic, i);
+
+	  fGenTypeMap[i + nNu] = "Cosmic";
+    
+	  const std::vector<art::Ptr<simb::MCParticle> > particles = truthCosmicMCPAssn.at(mcTruth.key());
+    
+	  for (auto const& particle : particles)
+	    {
+	      fTrackIDToGenMap[particle->TrackId()] = i + nNu;
+	    }
+	  ++nCos;
+	}
+      }
+  
+    eventID = e.event();
+    subRunID = e.subRun();
+    runID = e.run();
+  }
+
+  void CRUMBS::produce(art::Event& e)
+  {
+    if(fTrainingMode)
+      {
+	this->ClearMaps();
+	this->SetupMaps(e);
+      }
+
+
+    auto resultsVec = std::make_unique<std::vector<CRUMBSResult>>();
+    auto sliceAssns = std::make_unique<art::Assns<recob::Slice, CRUMBSResult>>();
+
+    auto const handleSlices(e.getValidHandle<std::vector<recob::Slice>>(fSliceModuleLabel));
+    std::vector<art::Ptr<recob::Slice>> slices;
+    art::fill_ptr_vector(slices, handleSlices);
+
+    auto const handlePFPs(e.getValidHandle<std::vector<recob::PFParticle>>(fPFParticleModuleLabel));
+    std::vector<art::Ptr<recob::PFParticle>> pfps;
+    art::fill_ptr_vector(pfps, handlePFPs);
+
+    auto const handleHits(e.getValidHandle<std::vector<recob::Hit>>(fHitModuleLabel));
+    std::vector<art::Ptr<recob::Hit>> allHits;
+    art::fill_ptr_vector(allHits, handleHits);
+
+    art::FindManyP<larpandoraobj::PFParticleMetadata> pfpMetadataAssoc(handlePFPs, e, fPFParticleModuleLabel);
+    art::FindManyP<sbn::SimpleFlashMatch> pfpFMAssoc(handlePFPs, e, fFlashMatchModuleLabel);
+
+    for(auto const &slice : slices)
+      {
+	this->ResetVars();
```

Comment:
Likewise, the vars should be scoped to within this loop.

Reply:
I think this is a required artefact of TMVA / TTree though right? These variables need to be members and so need resetting at the start of each unit, in this case a slice.